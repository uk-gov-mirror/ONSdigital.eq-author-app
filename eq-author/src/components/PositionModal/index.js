import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { uniqueId } from "lodash";
import styled from "styled-components";
import { InnerModal } from "./InnerModal";
import { Option } from "components/ItemSelectModal/ItemSelect";

const Indent = styled(Option)`
  margin-left: ${({ indent }) => (indent ? 1 : 0)}em;
`;

const PositionModal = ({ title, options, onMove, selected, onChange }) => {
  const positionButtonId = uniqueId("PositionModal");
  const [isOpen, setIsOpen] = useState(false);

  const previousIndex = options.findIndex(({ id }) => id === selected?.id);
  const previousPosition = useRef(previousIndex > -1 ? previousIndex : 0);

  const [{ position, item }, setOption] = useState({
    position: previousPosition.current,
    item: options[previousPosition.current],
  });

  useEffect(() => {
    // resets the position of the selected item when changing sections
    const previousIndex = options.findIndex(({ id }) => id === selected?.id);
    previousPosition.current = previousIndex > -1 ? previousIndex : 0;
    setOption((prev) => ({ ...prev, position: previousPosition.current }));
  }, [options, selected]);

  const orderedOptions = options.filter(({ id }) => id !== selected?.id);
  selected.parentEnabled = item?.parentEnabled;
  orderedOptions.splice(position, 0, selected);

  const handleClose = () => {
    setIsOpen(false);
    setOption({
      position: previousPosition.current,
      item: orderedOptions[previousPosition.current],
    });
  };

  const handleChange = (value) => {
    const count =
      orderedOptions[value]?.__typename === "Folder" && value - position >= 0
        ? orderedOptions.filter(
            ({ parentId }) => parentId === orderedOptions[value].id
          ).length
        : 0;
    setOption({
      position: parseInt(value, 10) + count,
      item: orderedOptions[value],
    });
  };

  const handleConfirm = (e) => {
    e.preventDefault();
    // item is the element in the modal you just clicked
    const { parentId = null } = item;

    let positionCalculation;

    if (parentId) {
      // get pages in target folder + selected item
      positionCalculation = orderedOptions
        .filter(
          ({ parentId: itemId, id }) =>
            parentId === itemId || id === selected.id
        )
        .findIndex(({ id }) => id === selected.id);
    } else {
      // remove all nested pages
      positionCalculation = orderedOptions
        .filter(({ parentId }) => !parentId)
        .findIndex(({ id }) => id === selected.id);
    }

    if (onMove) {
      onMove({
        ...item,
        folderId: parentId,
        position: positionCalculation,
      });
    }

    setIsOpen(false);
  };

  return (
    <div data-test="position-modal">
      <InnerModal
        id={positionButtonId}
        title={title}
        isOpen={isOpen}
        onClick={() => setIsOpen(true)}
        onClose={handleClose}
        onConfirm={handleConfirm}
        onChange={onChange || handleChange}
        selected={String(position)}
        displayName={selected.displayName}
      >
        {orderedOptions.map(({ displayName, parentEnabled }, i) => (
          <Indent
            data-test={`option-${i}`}
            key={i}
            value={String(i)}
            indent={parentEnabled ? parentEnabled.toString() : undefined}
          >
            {displayName}
          </Indent>
        ))}
      </InnerModal>
    </div>
  );
};

PositionModal.propTypes = {
  title: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      displayName: PropTypes.string.isRequired,
      folderId: PropTypes.string,
      parentEnabled: PropTypes.bool,
    })
  ).isRequired,
  onMove: PropTypes.func.isRequired,
  selected: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
    position: PropTypes.number,
  }).isRequired,
  onChange: PropTypes.string,
};

export default PositionModal;
