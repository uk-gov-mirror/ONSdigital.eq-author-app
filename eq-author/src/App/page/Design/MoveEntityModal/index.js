import React, { useMemo, useState } from "react";
import MoveModal from "components/MoveModal";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import PositionModal from "components/PositionModal";

import { useQuestionnaire } from "components/QuestionnaireContext";

export const buildPageList = (folders) => {
  const optionList = [];
  folders.forEach((folder) => {
    const { id, enabled, pages } = folder;
    if (enabled) {
      optionList.push({
        ...folder,
        parentEnabled: false,
      });
    }
    pages.forEach((page) => {
      optionList.push({
        ...page,
        parentId: enabled ? id : null,
        parentEnabled: enabled,
      });
    });
  });
  return optionList;
};

const propTypes = {
  sectionId: PropTypes.string.isRequired,
  page: CustomPropTypes.page,
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onMove: PropTypes.func.isRequired,
  entity: PropTypes.string.isRequired,
};

const entityData = {
  Folder: {
    options: (folders) =>
      folders.map((item) => (!item.enabled ? item.pages[0] : item)),
    move: ({ selected, sectionId, selectedSectionId, folderId, position }) => ({
      from: {
        id: selected.id,
        sectionId,
        position: selected.position,
      },
      to: {
        id: selected.id,
        selectedSectionId,
        folderId,
        position: position,
      },
    }),
  },
  Page: {
    options: (folders) => buildPageList(folders),
    move: ({ selected, sectionId, selectedSectionId, folderId, position }) => ({
      from: {
        id: selected.id,
        sectionId,
        position: selected.position,
      },
      to: {
        id: selected.id,
        sectionId: selectedSectionId,
        folderId,
        position: position,
      },
    }),
  },
};

const MoveEntityModal = ({
  sectionId,
  selected,
  isOpen,
  onClose,
  onMove,
  entity,
}) => {
  const { questionnaire } = useQuestionnaire();
  const { options, move } = entityData[entity];

  const [selectedSectionId, setSelectedSectionId] = useState(sectionId);

  const handleMove = (result) => {
    const { position, folderId } = result;
    onMove(
      move({ selected, sectionId, selectedSectionId, folderId, position })
    );
  };

  const selectedSection =
    questionnaire &&
    questionnaire.sections.find(({ id }) => id === selectedSectionId);

  return useMemo(
    () =>
      !questionnaire ? null : (
        <MoveModal
          title={entity === "Page" ? "Move question" : "Move folder"}
          isOpen={isOpen}
          onClose={onClose}
        >
          <PositionModal
            data-test={"section-position-modal"}
            title={"Section"}
            options={questionnaire.sections}
            selected={selectedSection}
            onChange={({ value }) =>
              setSelectedSectionId(questionnaire.sections[value].id)
            }
          />
          <PositionModal
            data-test={`${entity.toLowerCase()}-position-modal`}
            title={entity}
            options={
              selectedSection?.folders.length &&
              options(selectedSection.folders)
            }
            onMove={handleMove}
            selected={selected}
          />
        </MoveModal>
      ),
    [selectedSection, questionnaire, selected, isOpen, entity]
  );
};

MoveEntityModal.propTypes = propTypes;

export default MoveEntityModal;
