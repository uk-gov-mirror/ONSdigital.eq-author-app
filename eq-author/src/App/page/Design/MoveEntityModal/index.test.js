import React from "react";
import MoveEntityModal, { buildPageList } from ".";
import {
  render,
  screen,
  fireEvent,
  waitForElementToBeRemoved,
} from "tests/utils/rtl";
import { buildQuestionnaire } from "tests/utils/createMockQuestionnaire";
import { useQuestionnaire } from "components/QuestionnaireContext";

jest.mock("components/QuestionnaireContext", () => ({
  __esModule: true,
  useQuestionnaire: jest.fn(),
}));

const mockQuestionnaire = buildQuestionnaire({ sectionCount: 2 });
const currentSection = mockQuestionnaire.sections[0];
const currentPage = currentSection.folders[0].pages[0];

useQuestionnaire.mockImplementation(() => ({
  questionnaire: mockQuestionnaire,
}));
const onMove = jest.fn();
function setup(props) {
  const utils = render(
    <MoveEntityModal
      entity="Page"
      isOpen
      onClose={jest.fn()}
      sectionId={currentSection.id}
      selected={currentPage}
      onMove={onMove}
      {...props}
    />
  );
  return { ...utils, onMove };
}

describe("MovePageModal", () => {
  beforeEach(() => setup());

  it("should render rtl", () => {
    expect(screen.getByTestId("move-modal")).toBeVisible();
  });

  it("should open section select modal", () => {
    expect(screen.queryByTestId("section-item-select")).toBeNull();
    fireEvent.click(screen.getByText(/Section 1/));
    expect(screen.queryByTestId("section-item-select")).toBeVisible();
  });

  it("should close select modal on select", async () => {
    fireEvent.click(screen.getByText(/Section 1/));

    fireEvent.click(screen.getByText(/Select/));

    await waitForElementToBeRemoved(() => screen.queryByText(/Select/i));
    expect(screen.queryByText(/Select/)).toBeNull();
  });

  it("should close select modal on cancel", async () => {
    fireEvent.click(screen.getByText(/Section 1/));

    fireEvent.click(screen.getByText(/Cancel/));

    await waitForElementToBeRemoved(() => screen.queryByText(/Cancel/i));
    expect(screen.queryByText(/Cancel/)).toBeNull();
  });

  it("should update selection on change", async () => {
    fireEvent.click(screen.getByText(/Section 1/));

    fireEvent.click(screen.getByText(/Section 2/));
    fireEvent.click(screen.getByText(/Select/));
    await waitForElementToBeRemoved(() => screen.queryByText(/Select/i));
    expect(screen.queryByText(/Section 1/)).toBeNull();
    expect(screen.queryByText(/Section 2/)).toBeVisible();
  });
});

describe("MovePageModal: questionnaire not loaded", () => {
  it("shouldn't render if questionnaire not yet available", () => {
    useQuestionnaire.mockImplementation(() => ({
      questionnaire: undefined,
    }));

    setup({ selected: null });

    expect(screen.queryByTestId("move-modal")).toBeNull();
  });
});
describe("MovePageModal: buildPageList", () => {
  let output, folders;

  beforeEach(() => {
    folders = [
      {
        id: "folder-1",
        enabled: true,
        position: 0,
        __typename: "Folder",
        pages: [
          {
            id: "folder-1-page-1",
            position: 0,
          },
          {
            id: "folder-1-page-2",
            position: 1,
          },
        ],
      },
      {
        id: "folder-2",
        enabled: false,
        position: 1,
        pages: [
          {
            id: "folder-2-page-1",
            position: 2,
          },
        ],
      },
    ];
    output = buildPageList(folders);
  });

  it("should include enabled folders in output", () => {
    expect(
      output.filter(({ __typename }) => __typename === "Folder")[0]
    ).toMatchObject({
      __typename: "Folder",
      parentEnabled: false,
      enabled: true,
    });
  });

  it("should not included disabled folders in output", () => {
    expect(
      output.filter(({ enabled, type }) => !enabled && type === "Folder")
    ).toHaveLength(0);
  });

  it("should include parent details on enabled folders", () => {
    expect(output.filter(({ id }) => id.startsWith("folder-1-page"))).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          parentEnabled: true,
          parentId: output[0].id,
          position: expect.any(Number),
        }),
      ])
    );
  });

  it("should have null parentId for disabled folders", () => {
    const [target] = output.filter(({ id }) => id.includes("folder-2"));
    expect(target.parentId).toBeNull();
  });
});
