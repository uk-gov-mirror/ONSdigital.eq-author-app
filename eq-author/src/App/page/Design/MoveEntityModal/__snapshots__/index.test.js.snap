// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovePageModal should render 1`] = `undefined`;
